### DISCUSSION TITLE
Trailing Slash Usage in the MindConnect NodJS

### DISCUSSION THREAD

**Discussion by tntstn** [mindsphere/mindconnect-nodejs#355] [General]
I have a question and need an answer immediately. There is a restriction which is defined in insightshub (Insights Hub Release Notes mention change in AssetManager api behaviour) release document. In this document, basicly it says "If there is any usage at the end of the requests with backslash / , it should be removed ". As you see below, there is no backslash at the end of the request explained under well-formed requests.

Currently, we are using "@mindconnect/mindconnect-nodejs": "3.22.0". I need to know, is there any usage in this version or if there is an usage in any version, which version is the proper to upgrade?

Examples of malformed requests:
GET [https://gateway.\Unknown](https://gateway.%5Cunknown/) macro: {region}-{environment}.{domain}/api/.../v3/.../{Id}/{Name}/

Examples of well-formed requests:
GET [https://gateway.\Unknown](https://gateway.%5Cunknown/) macro: {region}-{environment}.{domain}/api/.../v3/.../{Id}/{Name}

**Comment by sn0wcat** [comment]
@tntstn as far as i can remeber, this trailing slash issue is known since 3-4 years, and i have looked at that time if there are any malformed requests or not - I don't think that there are any, but if you encounter some, please tell

**Comment by mervoli** [comment]
Hi all, When I checked the source code (3.23.0 - latest) manually, I encountered it in three places, I am not sure if they are valid, please correct me if I'm wrong.
1) In "api/sdk/asset/asset-management.ts" (Line1307): GetBillboard()
2) In "api/sdk/event/event-management.ts" (Line 21): GetBillboard()
3) In "api/sdk/open-edge/open-edge.ts" (Line 548): PostDeviceHeartbeat()

**Comment by sn0wcat** [comment]
we could add in the implementations of HttpAction a check which just removes the trailing slash if there is one.

https://github.com/mindsphere/mindconnect-nodejs/blob/bca21837551535b20fac5632bca247841e9ca59d/src/api/frontend-auth.ts#L112

https://github.com/mindsphere/mindconnect-nodejs/blob/bca21837551535b20fac5632bca247841e9ca59d/src/api/browser-auth.ts#L155

```code
url.replace(/\/$/, '')
```

**Reply by mervoli** [reply]
sounds like a very good solution.

**Reply by tntstn** [reply]
@sn0wcat when you publish the new version could you ping us? Then we will update our dependency.

**Comment by sn0wcat** [comment]
The billboard urls are correct btw - take a look here - it seems that the insights hub rules about trailing slash have not been verified against insights hub own APIs ;) cc: @gesslinger 

https://developer.siemens.com/insights-hub/docs/apis/advanced-assetmanagement/api-assetmanagement-api.html

![image](https://github.com/mindsphere/mindconnect-nodejs/assets/1426617/408bbaca-d95b-4674-8f8e-187f3abb0a5e)

**Reply by sn0wcat** [reply]
@tntstn @mervoli there is now a v 3.24.2 which automatically removes all slashes from the URL except for the two Billboard URLs in EventManagement and AssetManagement, otherwise these two endpoints would stop working

https://www.npmjs.com/package/@mindconnect/mindconnect-nodejs/v/3.24.2

**Reply by tntstn** [reply]
Maybe this page was not updated yet. But it's explained in Release Notes 23 march 2024 under Asset Management section: https://documentation.mindsphere.io/MindSphere/apps/release-notes/march-2024.html
@sn0wcat @gesslinger
